### DISCUSSION TITLE
How get Session in console?

### DISCUSSION THREAD

**Discussion by genby8** [symfony/symfony#42461] [Q&A]
I use Symfony 5.3.6 PHP 7.4.14
i inject $requestStack in constructor my command class.

in method execute i call $requestStack->getCurrentRequest()->getSession(), but getCurrentRequest() return null.
How get session? SessionInterface is deprecated

**Comment by javiereguiluz** [comment]
In Console commands there's no HTTP context, so there are no requests and no sessions. It's mentioned here: https://symfony.com/doc/current/routing.html#generating-urls-in-commands

If you want to control the browser programmatically with a console command, see https://github.com/symfony/panther

**Reply by genby8** [reply]
https://symfony.com/blog/new-in-symfony-5-3-session-service-deprecation 

> Sessions are not part of the HTTP specification (either HTTP/1.1, HTTP/2 or HTTP/3) because HTTP is stateless. That’s why it feels odd to handle sessions as part of the HttpFoundation component.

(sorry for my bad english)

Session is indirectly related to HTTP Request. Why there are no requests in requestStack I can still understand. But now it turns out the session is bound to the HTTP Request. Weird.

**Reply by dbrumann** [reply]
What do you need the session for in your command? Maybe something else fits better with your CLI requirements, e.g. a database or cache?

**Reply by genby8** [reply]
(sorry for my bad english)

Unfortunately, I have a legacy project starting in Symfony on Zend 1, which is configured so that the session is used very deeply in models.

I understand perfectly well that the session is used for other purposes, but I want the control framework to start the session. And now I don't understand how to start it correctly: by yourself or still get it from the service? But how?

**Reply by dbrumann** [reply]
If you really need a request in your CLI application, you probably want to "simulate" a request by creating it yourself and then pushing it to the request stack early on in your console application, e.g. something like this:

```
$session = new Session(); // You might want to use an alternative $storage instead of the default native storage
$request = new Request(); // You might want set some details like url or method based on your needs
$request->setSession($session);

$this->requestStack->push($request);
```

Now, whenever your code wants to get a request from the stack it will get your artificial request with the custom session.

Be aware, that this might have unwanted side effects and cause issues in weird places, e.g. wherever additional details are needed on the request. For example in a regular HTTP request Symfony will add attributes - e.g. `_route` to see which route a request matches -, which will not be available in this example above. In other words, you might run into some unexpected code paths because your request will not be a "real" HTTP request. So, this is far from ideal.

**Reply by genby8** [reply]
`$this->requestStack->push($request);`
Here in this method "push" an event can be triggered to process this request (although it is not there now), so this solution is not good.
`$session = new Session()`
I also want to take this already configured from the configuration.
We'll have to clean the Legacy. But all the same it is interesting why the creation of the session was delegated to the request.

Thank you for your help.

**Reply by stof** [reply]
The request stack is not responsible for triggering events. That's the HttpKernel doing it (which is also the one pushing the "normal" request in the stack)

**Reply by genby8** [reply]
(sorry for my bad english)
The push method in RequestStack is public. I see that RequestStack resolves a lot. Pop and Push methods are superfluous here. So you either remove them in the next releases (let's say you still make the storage read-only), or hang up the event call. I really hope for the first option.
But it is still not clear what to do with the session and why you are inviting it to the request. Or do you plan to do your session at the subquery?

**Reply by linaori** [reply]
To get the session and have all listeners function you could try the following (taken from `HttpKernel::handleRaw()`)
```php
$request = new Request();
$this->requestStack->push($request);

$event = new RequestEvent($httpKernel, $request, HttpKernelInterface::MAIN_REQUEST);
$this->dispatcher->dispatch($event, KernelEvents::REQUEST);
```
Original:
https://github.com/symfony/symfony/blob/e40e29339d632d5ff48fe4f4edac4397f9cf083d/src/Symfony/Component/HttpKernel/HttpKernel.php#L126-L132

**Reply by genby8** [reply]
Thanks for the advice. But doesn't it seem like duplication of code? Copying a piece of code from a version of a package is wrong in every way.

Symfony has wrapped container handling using autowiring. Now it is reasonable to prohibit making public services. Therefore, you need functionality out of the box, and not initiate the main request somewhere in a self-written service.
If I use the code described above in the service, then the firewall will fire (and anything can happen here). If I use subqueries without the main query, then this is strange.

I came to the conclusion that there is no such functionality at the current time (v. 5.3.x and higher).

**Reply by linaori** [reply]
> Thanks for the advice. But doesn't it seem like duplication of code? Copying a piece of code from a version of a package is wrong in every way.

You can create an event listener or subscriber that sets this for every command, or based on a criteria for a command. See: https://symfony.com/doc/current/components/console/events.html

The problem here is not that you have to copy-paste the code, the problem is that you're trying to use something designed to be used in the context of webserver requests in an environment where they don't exist.

> If I use the code described above in the service, then the firewall will fire (and anything can happen here). If I use subqueries without the main query, then this is strange.

Calling command code from a web request is not recommended either. If you want to share code between the web and cli, move the code to a shared dependency. These dependencies should not depend on the http-foundation and console components.

**Reply by genby8** [reply]
> Проблема здесь не в том, что вам нужно копировать-вставить код, проблема в том, что вы пытаетесь использовать что-то, предназначенное для использования в контексте запросов веб-сервера в среде, где они не существуют.

We're talking about session and not about cookies, right?

**Reply by linaori** [reply]
> We're talking about session and not about cookies, right?

`Symfony\Component\HttpFoundation\Session\Session` which is from the http-foundation. While sessions use a cookie to identify the session id from the client, sessions are designed to carry information across web requests regardless of how the session id is stored.

You can make it work on the cli, but this is "use at your own risk".
